Framework/Algorithms/src/CalculatePolynomialBackground.cpp

```diff
@@ -196,41 +206,66 @@ executeFit(Mantid::API::Algorithm &fit, const std::string &function,
 }
 
 /** Return a Fit algorithm compatible string representing a polynomial.
+ *  @param name a string respresenting the name of the polynomial
  *  @param parameters a vector containing the polynomial coefficients
  *  @return a function string
  */
-std::string makeFunctionString(const std::vector<double> &parameters) {
+std::string makeFunctionString(const std::string &name,
+                               const std::vector<double> &parameters) {
   const auto degree = parameters.size() - 1;
   std::ostringstream s;
+  s << "name=" << name;
+  if (degree > 2)
+    s << ",n=" << degree;
+  for (size_t d = 0; d <= degree; ++d) {
+    s << ',' << 'A' << d << '=' << parameters[d];
+  }
+  return s.str();
+}
+
+/** Return a name of the function used in the fit.
+ *  @param degree an integer representing the degree of the polynomial
+ *  @return a string containing the name of the polynomial
+ */
+std::string makeNameString(const size_t degree) {
+  std::ostringstream name;
   switch (degree) {
   case 0:
-    s << "name=FlatBackground";
+    name << "FlatBackground";
     break;
   case 1:
-    s << "name=LinearBackground";
+    name << "LinearBackground";
     break;
   case 2:
-    s << "name=Quadratic";
+    name << "Quadratic";
     break;
   default:
-    s << "name=Polynomial,n=" << degree;
+    name << "Polynomial";
   }
-  for (size_t d = 0; d <= degree; ++d) {
-    s << ',' << 'A' << d << '=' << parameters[d];
-  }
-  return s.str();
+  return name.str();
 }
 
 /** Evaluates the given function directly on a histogram
- *  @param function a string representing function to evaluate
+ *  @param name a string representing the name of the polynomial to evaluate
+ *  @param parameters a vector containing the coefficients of the polynomial
  *  @param ws an output workspace
  *  @param wsIndex a workspace index identifying a histogram
  */
-void evaluateInPlace(const std::string &function,
+void evaluateInPlace(const std::string &name,
+                     const std::vector<double> &parameters,
                      Mantid::API::MatrixWorkspace &ws, const size_t wsIndex) {
+  const auto degree = parameters.size() - 1;
   auto bkg = boost::dynamic_pointer_cast<Mantid::API::IFunction1D>(
-      Mantid::API::FunctionFactory::Instance().createInitialized(function));
-  // We want to write directly to the workspace.
+      Mantid::API::FunctionFactory::Instance().createFunction(name));
+  if (degree > 2) {
+    Mantid::API::IFunction1D::Attribute att = bkg->getAttribute("n");
+    att.fromString(std::to_string(degree));
+    bkg->setAttribute("n", att);
+  }
+  for (size_t d = 0; d <= degree; ++d) {
+    std::string param = 'A' + std::to_string(d);
+    bkg->setParameter(param, parameters[d]);
```

Comment:
I'm trying to wrap my head around why IN5 system test needed to be updated.
Can you confirm that it's this block that is changing the behaviour also for 0-degree fitting?

Reply:
Yes this block is different for all degrees of polynomial. 

The differences between the new output and the original reference for the system test seems to be quite small. I have just tried setting the relative error to false in the system test there is then only one mismatch of data: 
``` 
Data mismatch at cell (hist#,bin#): (267,179)
 Dataset #1 (X,Y,E) = (2.023,4.68025,0.273321)
 Dataset #2 (X,Y,E) = (2.023,4.68025,0.273322)
 Difference (X,Y,E) = (0,1.18704e-07,1.27676e-07)
```